Add automatic responses for OPTIONS, invalid methods, and failed content negotiation

yunyu opened the pull request:
Also implements a internal framework for adding other automatic responses.

* Does nothing if a CorsHandler an OPTIONS method is added. 
* 405 if request specifies a wrong verb
* 415 if consumed content negotiation fails
* 406 if produced content negotiation fails

Fixes #594 and #456.

I don't like the instanceof check, but it's a bit too late to force users to declare `.method(HttpMethod.OPTIONS)` when adding a CorsHandler. The only other solution is to do a massive refactor of response handling to allow replacing the response after it's written (which is unlikely to happen).

This will break response handling for the very small (probably 0) minority of users who handle OPTIONS without declaring the method in the route or using CorsHandler, as well as for people who expect non-compliant behavior for the other fixes.

pmlopes commented:
This is part of the desired behavior, we could either accept it and later improve though. What is missing is that a OPTIONS response should take all routes in consideration to identify the available methods, plus if in a route there is a GET but no HEAD, HEAD should be added to the list probably and also handled since it should behave like a GET but terminate at end of writting headers...

yunyu commented:
I don't think there's a hurry at all, we can look more deeply in the spec (and use the framework to fix #456 as well).

yunyu commented:
@pmlopes Added an AutomaticHandler that handles failed content negotiation. What do you think?

yunyu commented:
Just kidding, implemented the correct behavior for 415, 406, 405, OPTIONS. Some of the checks in `Route.matches()` can be optimized out, but I think that's just a map/collect away.

yunyu commented:
Implemented appropriate responses for all cases other than the HEAD/GET stuff (I think a stub HEAD handler is possible only if we can provide a proxied HttpServerResponse that refuses to send any body) and added tests. Also optimized Route##match to skip parameter calculations if content type negotiation fails in either direction (not sure why this wasn't the case before).

slinkydeveloper commented:
I close as it's too old and invalid and partially replaced by #1305
